controllers/openstackdataplanenode_controller.go

```diff
@@ -226,10 +241,18 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
 		instance.Status.Conditions.Set(condition.TrueCondition(condition.DeploymentReadyCondition, condition.DeploymentReadyMessage))
 
 		// Explicitly set instance.Spec.Deploy = false
+		// Explicitly set instance.Spec.DeployStrategy.DeployIdentifier = ""
 		// We don't want another deploy triggered by any reconcile request, it
 		// should only be triggered when the user (or another controller)
 		// specifically sets it to true.
+		// patch := client.MergeFrom(instance.DeepCopy())
+		r.Log.Info("Set DeployStrategy.Deploy to false")
 		instance.Spec.DeployStrategy.Deploy = false
+		instance.Spec.DeployStrategy.DeployIdentifier = ""
```

Comment:
This ties into my comments above. We end up setting a new string for the deployIdentifier above, then reset it to "" here?. So when it gets back to the above if statement, it's `== 0` and off we go again.

We will need to make sure we maintain logic that will prevent another rollout when we update this. The controller will still reconcile when we do this, so we'll need to have the logic in place to handle it.

Reply:
This line is never reached until the deployment is done. If a deployment is in progress, we've returned earlier. By the time we reach here, the deployment is done and we reset Deploy and DeployIdentifier back to the non-deploying values. On the next (auto)reconcile a deploy is not started since Deploy=false.

Reply:
As I said earlier, what @slagle says here is correct. Those values are set to their default only when the deploy is done.

Again, the fact that it starts from scratch after some minutes (?) it doesn't caused by this (IMHO).

Reply:
> This line is never reached until the deployment is done. If a deployment is in progress, we've returned earlier. By the time we reach here, the deployment is done and we reset Deploy and DeployIdentifier back to the non-deploying values. On the next (auto)reconcile a deploy is not started since Deploy=false.

It definitely is:
```
{"level":"info","ts":"2023-05-31T20:52:59+10:00","msg":"Adding label openstackdataplanerole=edpm-compute","controller":"openstackdataplanenode","controllerGroup":"dataplane.openstack.org","controllerKind":"OpenStackDataPlaneNode","OpenStackDataPlaneNode":{"name":"edpm-compute-0","namespace":"openstack"},"namespace":"openstack","name":"edpm-compute-0","reconcileID":"9dc6c525-9c9c-4df8-a424-cb705cb39c1e"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","msg":"Node","controller":"openstackdataplanenode","controllerGroup":"dataplane.openstack.org","controllerKind":"OpenStackDataPlaneNode","OpenStackDataPlaneNode":{"name":"edpm-compute-0","namespace":"openstack"},"namespace":"openstack","name":"edpm-compute-0","reconcileID":"9dc6c525-9c9c-4df8-a424-cb705cb39c1e","DeployStrategy":true,"Node.Namespace":"openstack","Node.Name":"edpm-compute-0"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","logger":"controllers.OpenStackDataPlane","msg":"Create Or Patch DataPlaneRole","roleName":"edpm-compute"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","logger":"controllers.OpenStackDataPlane","msg":"CreateDataPlaneNode","nodeName":"edpm-compute-0"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","logger":"controllers.OpenStackDataPlane","msg":"Create Or Patch DataPlaneRole","roleName":"edpm-compute"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","logger":"controllers.OpenStackDataPlane","msg":"Set DeployStrategy.Deploy to false"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","msg":"DeployIdentifier updated to: ","controller":"openstackdataplanenode","controllerGroup":"dataplane.openstack.org","controllerKind":"OpenStackDataPlaneNode","OpenStackDataPlaneNode":{"name":"edpm-compute-0","namespace":"openstack"},"namespace":"openstack","name":"edpm-compute-0","reconcileID":"9dc6c525-9c9c-4df8-a424-cb705cb39c1e","identifier":"672c04ae-9681-48fa-b7de-96bfb939b03b"}
{"level":"info","ts":"2023-05-31T20:52:59+10:00","msg":"Starting DataPlaneNode deploy","controller":"openstackdataplanenode","controllerGroup":"dataplane.openstack.org","controllerKind":"OpenStackDataPlaneNode","OpenStackDataPlaneNode":{"name":"edpm-compute-0","namespace":"openstack"},"namespace":"openstack","name":"edpm-compute-0","reconcileID":"9dc6c525-9c9c-4df8-a424-cb705cb39c1e"}
```

We can see there that it clears the `deployIdentifer` and starts a new reconcile since the reconcileID has changed. This loops on repeat indefinitely. So we start reconciling, then we update the Spec with the `deployIdetifier` UUID, we update the Spec with the empty value for `deployIdentifier`, then we start again.

If you `make run`, create a dataplanenode and then just run this a few times in succession:
```
❯ oc get openstackdataplanenode edpm-compute-0 -o jsonpath='{.spec.deployStrategy}' | jq .
```

You can see that it does exactly there, where it's adding, clearing, re-adding that `deployIdentifer`.

Reply:
That would be the issue then if deploy is not getting set back to false.

Reply:
I don't see the behavior you describe. How exactly did you reproduce it?

This is what I did:

I have a diff to disable the actual deployment execution for a node, and it adds a 5 sec sleep. This gives me time to query the node and check the deployStrategy values:

```
diff --git a/controllers/openstackdataplanenode_controller.go b/controllers/openstackdataplanenode_controller.go
index 40c383c..425ee30 100644
--- a/controllers/openstackdataplanenode_controller.go
+++ b/controllers/openstackdataplanenode_controller.go
@@ -191,7 +191,7 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
 		}
 	}
 
-	nodeConfigMap, err := r.GenerateInventory(ctx, instance, instanceRole)
+	// nodeConfigMap, err := r.GenerateInventory(ctx, instance, instanceRole)
 	if err != nil {
 		util.LogErrorForObject(helper, err, fmt.Sprintf("Unable to generate inventory for %s", instance.Name), instance)
 		return ctrl.Result{}, err
@@ -204,6 +204,7 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
 		if len(instance.Spec.DeployStrategy.DeployIdentifier) == 0 {
 			// patch := client.MergeFrom(instance.DeepCopy())
 			newIdentifier := dataplanev1beta1.GenerateDeployIdentifier()
+			r.Log.Info("#############GENERATING DEPLOYIDENTIFIER", "newIdentifier", newIdentifier)
 			instance.Spec.DeployStrategy.DeployIdentifier = newIdentifier
 			err = r.Client.Update(ctx, instance)
 			if err != nil {
@@ -216,11 +217,13 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
 		r.Log.Info("Set DeploymentReadyCondition false", "instance", instance)
 		instance.Status.Conditions.Set(condition.FalseCondition(condition.DeploymentReadyCondition, condition.RequestedReason, condition.SeverityInfo, condition.DeploymentReadyRunningMessage))
 
-		nodes := &dataplanev1beta1.OpenStackDataPlaneNodeList{
-			Items: []dataplanev1beta1.OpenStackDataPlaneNode{*instance},
-		}
+		// nodes := &dataplanev1beta1.OpenStackDataPlaneNodeList{
+		// 	Items: []dataplanev1beta1.OpenStackDataPlaneNode{*instance},
+		// }
 
-		deployResult, err := deployment.Deploy(ctx, helper, instance, nodes, ansibleSSHPrivateKeySecret, nodeConfigMap, &instance.Status, instance.GetAnsibleEESpec(*instanceRole), r.GetServices(instance, instanceRole), instanceRole)
+		// deployResult, err := deployment.Deploy(ctx, helper, instance, nodes, ansibleSSHPrivateKeySecret, nodeConfigMap, &instance.Status, instance.GetAnsibleEESpec(*instanceRole), r.GetServices(instance, instanceRole), instanceRole)
+		r.Log.Info("##############SLEEPING FOR 5 SECS")
+		time.Sleep(5 * time.Second)
 		if err != nil {
 			util.LogErrorForObject(helper, err, fmt.Sprintf("Unable to deploy %s", instance.Name), instance)
 			instance.Status.Conditions.Set(condition.FalseCondition(
@@ -231,10 +234,10 @@ func (r *OpenStackDataPlaneNodeReconciler) Reconcile(ctx context.Context, req ct
 				err.Error()))
 			return ctrl.Result{}, err
 		}
-		if deployResult != nil {
-			result = *deployResult
-			return result, nil
-		}
+		// if deployResult != nil {
+		// 	result = *deployResult
+		// 	return result, nil
+		// }
 
 		instance.Status.Deployed = true
 		r.Log.Info("Set DeploymentReadyCondition true", "instance", instance)
```

I make sure CRD's are updated in the cluster, start the operator (go run main.go), create the sample edpm-compute-0 (with the role line removed), patch the node to set deployStrategy.deploy to true, then check the deployStrategy values.

```
oc create -f config/samples/dataplane_v1beta1_openstackdataplanenode_edpm_compute_0.yaml
oc patch openstackdataplanenode edpm-compute-0  -p='[{"op": "replace", "path": "/spec/deployStrategy/deploy", "value":true}]' --type json
oc get osdpnode edpm-compute-0 -o json | jq .spec.deployStrategy
```

Once I hit the sleep, I get the following:

```
{
  "deploy": true,
  "deployIdentifier": "406efbab-6fda-4b3b-8e75-ab41ab4a7979"
}
```

Once past the sleep, deploy is toggled back to false and deployIdentifier is dropped (I suppose due to it being set back to ""). I never see a new value generated.

```
{
  "deploy": false
}
```

To simulate an actual deployment (with requeues as if it was watching for changes to an OpenStackAnsibleEE), I add the following line after where I've commented out the check for deployResult:

```
// if deployResult != nil {                           
//  result = *deployResult                            
//  return result, nil                                
// }                                                  
return ctrl.Result{RequeueAfter: time.Second * 2}, err
```

In that case, I see a single unique value for deployIdentifier generated and it stays consistent with that value (not changing).

However, if the node is owned by the dataplane, then setting deploy=true on the node and then seeing deployIdentifier constantly getting reset, I could see that happening due to the dataplane reconcile overwriting changes done to its owned node resource. Could that explain the behavior you saw?

Reply:
I can see a situation where if deploying from the DataPlane, the role could get suck in a loop. I'll comment in the code where I think this is happening.